action/protocol/staking/handler_candidate_endorsement.go

```diff
@@ -0,0 +1,103 @@
+package staking
+
+import (
+	"context"
+
+	"github.com/iotexproject/iotex-address/address"
+
+	"github.com/iotexproject/iotex-core/action"
+	"github.com/iotexproject/iotex-core/action/protocol"
+	"github.com/iotexproject/iotex-core/pkg/util/byteutil"
+)
+
+const (
+	handleCandidateEndorsement = "candidateEndorsement"
+)
+
+func (p *Protocol) handleCandidateEndorsement(ctx context.Context, act *action.CandidateEndorsement, csm CandidateStateManager) (*receiptLog, []*action.TransactionLog, error) {
+	actCtx := protocol.MustGetActionCtx(ctx)
+	featureCtx := protocol.MustGetFeatureCtx(ctx)
+	log := newReceiptLog(p.addr.String(), handleCandidateEndorsement, featureCtx.NewStakingReceiptFormat)
+
+	esm := NewEndorsementStateManager(csm.SM())
+	bucket, rErr := p.fetchBucket(csm, act.BucketIndex())
+	if rErr != nil {
+		return log, nil, rErr
+	}
+	cand := csm.GetByOwner(bucket.Candidate)
+	if cand == nil {
+		return log, nil, errCandNotExist
+	}
+	log.AddTopics(byteutil.Uint64ToBytesBigEndian(bucket.Index), bucket.Candidate.Bytes(), []byte{byteutil.BoolToByte(act.Endorse())})
+
+	var err error
+	if act.Endorse() {
+		err = p.endorseCandidate(ctx, csm, esm, actCtx.Caller, bucket, cand)
+	} else {
+		err = p.unEndorseCandidate(ctx, csm, esm, actCtx.Caller, bucket, cand)
+	}
+	if err != nil {
+		return log, nil, err
+	}
+	return log, nil, nil
+}
+
+func (p *Protocol) endorseCandidate(ctx context.Context, csm CandidateStateManager, esm *EndorsementStateManager, caller address.Address, bucket *VoteBucket, cand *Candidate) error {
+	if err := p.validateEndorse(ctx, csm, esm, caller, bucket, cand); err != nil {
+		return err
+	}
+
+	if err := esm.Put(bucket.Index, &Endorsement{
+		ExpireHeight: endorsementNotExpireHeight,
+	}); err != nil {
+		return csmErrorToHandleError(caller.String(), err)
```

Comment:
this is for csm to use? i. e., when the value/type of `err` is related to candidate error

Reply:
i mean, it will return a regular `err`, not the `handleError` you may want

Reply:
should return regular err here, b/c the error returned must be some unknown errors from db